src/mpid/common/genq/mpidu_genq_private_pool.c

```diff
@@ -190,10 +248,23 @@ int MPIDU_genq_private_pool_free_cell(MPIDU_genq_private_pool_t pool, void *cell
     }
 
     cell_h = CELL_TO_CELL_HEADER(cell);
+    block = cell_h->block;
+
+    cell_h->next = block->free_list_head;
+    block->free_list_head = cell_h;
 
-    cell_h->next = pool_obj->free_list_head;
-    pool_obj->free_list_head = cell_h;
-    cell_h->block->num_used_cells--;
+    block->num_used_cells--;
+
+    if (block->num_used_cells == pool_obj->num_cells_in_block - 1) {
+        append_free_blocks(pool_obj, block);
+    } else if (block->num_used_cells == 0) {
+        if (pool_obj->num_blocks > 1) {
```

Comment:
Can we a comment about the last block is will not be freed until the pool is destroyed?

Reply:
Or making it configurable?

Reply:
I'll add a comment. Let's avoid the extra configuring complexity until we have a specific reason.